### ISSUE TITLE
Globalmenu: dbusmenu-object on Dbus does not contain full layout

### ISSUE THREAD

**Issue by 1931186** [electron/electron#8297]
* Electron version: 1.4.13
* Operating system: Arch Linux

**TLDR**
Menuobject on electrons bus does not provide full layout. Not quite sure if bug or "feature".

I'm using this Appmenu-Registrar implementation:
https://github.com/RafaelBocquet/i3-hud-menu
to simulate Unitys com.canonical.AppMenu.Registrar dbus service which Unity uses for its global Menu.

### Expected behavior
The menuobject advertised on the Bus for this application:
/usr/lib/electron/electron --app=/usr/lib/atom --executed-from=$HOME --pid=$PID
(with $HOME being my home directory and $PID the process ID)
implements the com.canonical.dbusmenu interface and got registered at my mock-Registrar.
The advertised menuobject provides a GetLayout() function, that should return the menu layout.
Calling GetLayout(0, -1, ["label"]) should return the complete layout (0=parent ID=root of menu; -1=recursion depth=infinite; ["label"]=property names=return labels). Through testing normal gtk applications with the Unity-gtk-module I can confirm it normally does.

### Actual behavior
The menuobject doesn't contain the full layout, but only a "Root" and the 7 toplevel options "_Selection", "_View", "_Edit", "F_ind", "_Packages", "_File", "_Help".
The "Root" gets set here:
https://github.com/electron/electron/blob/787bc8570382e98c4f204abff05b2af122e5a422/atom/browser/ui/views/global_menu_bar_x11.cc#L191

### How to reproduce
Run on Linux distribution not using Unity and make sure you've got libdbusmenu-glib.so

I am quite certain this a bug because electron registers itself at the Registrar with this menu, and it wouldn't make much sense to then provide an incomplete layout. However it is also a bit unusual to create a "Root" item (you would have only one item in your menubar named "Root" with all the other entries being in its submenu), so I am not quite sure what to think of this.

**Comment by 1931186** [comment]
**Closed**
Problem was caused by the appmenu implementation I used. Now fixed.

**Comment by snippins** [comment]
@1931186 How do you fixed it? Thank you.

**Comment by 1931186** [comment]
@snippins The menu is a tree structure and the nodes are by default collapsed. You need to recursively call AboutToShow(item_id) + update the layout on your side via GetLayout() to expand it. The pygobject (Canonicals global menu implementation uses GLib) documentation is unfortunately pretty sparse so I had to go through the electron source code to realize this (you can find a list of the different item props in the file I've linked). 
Pseudocode:
```
expand(item):
    if [item has children]:
        call AboutToShow(item_id)
        call GetLayout()
        for children in item: #You need to find item again, my code is structured a bit differently but essentially you can use GetLayouts first parameter for this
            expand(child)
```
Also if you are familiar with canonicals global menu/ know a STABLE client for the unity-gtk-module I would be extremely thankful if you could point me to some info becuase I've implemented [this API](https://wiki.gnome.org/Projects/GLib/GApplication/DBusAPI) and it works but is super unstable (I've just tried [this](https://stackoverflow.com/questions/44143643) but it doesn't seem to work :()

**Comment by snippins** [comment]
Thanks for the info, the builtin KDE daemon does call AboutToShow() and does not have this problem. So I think you are right. 

Also, the second link you provided is for DBusMenu, thus it is not related to unity-gtk-module (I think). 

This project [vala-panel-appmenu](https://github.com/rilian-la-te/vala-panel-appmenu) implement an good client for unity-gtk-module (org.gtk.Menus). It use a [submodule](https://github.com/rilian-la-te/vala-dbusmenu) library, which you should look closely into.

It vala-appmenu project also provided and implementation for the Registrar interface, but have the same problem with i3-hud-service.py. Me and the author are also working on expose the full layout for electron apps and non-qt apps that use DBusMenu, you can track these bugs [1](https://github.com/rilian-la-te/vala-panel-appmenu/issues/89#issuecomment-303785657), [2](https://github.com/rilian-la-te/vala-dbusmenu/issues/6).

**Comment by 1931186** [comment]
Thanks for the link.

The unity-gtk-module [does use](http://bazaar.launchpad.net/~indicator-applet-developers/unity-gtk-module/trunk.15.04/view/head:/src/main.c) Gios g_dbus_connection_export_menu_model which returns a [GDBusMenuModel](https://www.freedesktop.org/software/gstreamer-sdk/data/docs/latest/gio/GDBusMenuModel.html) which is the thing I refer to in my second link but which also is completely separate from the DBusMenu from libdbusmenu-glib (which is the one used by electron).

I have a (working) python script that exports the full layout of both gtk-applications and applications that use libdbusmenu-glib (and qt applications to a certain extent) to a json file (by just using their respective dbus interfaces). My problem is that some gtk-applications (gimp) tend to crash (sporadically) when I Activate() an entry, and I tried to fix this by not using the ghetto method of exposing the whole menu on the bus but actually using Gios way of retrieving a menu from the bus and casting it to a gtk.Menu object. I will look into how vala-panel does this.

The bugreports you linked indicate that you think the problem lies at the Registrar which is not really the case (I also use i3-hud-service.py). What you have to do is expand the menu shortly before processing it, because the glib menu is mutable and can change 'on-the-fly'.

**Comment by rilian-la-te** [comment]
[vala-dbusmenu](https://github.com/rilian-la-te/vala-dbusmenu) is for `com.canonical.dbusmenu`, not for `org.gtk.Menus`))) To be correct:)

**Comment by 1931186** [comment]
@rilian-la-te how do you implement gtk menus then?

**Comment by rilian-la-te** [comment]
Look [here](https://github.com/rilian-la-te/vala-panel-appmenu/blob/master/lib/menu-widget-menumodel.vala).
Using builtin Gtk interfaces + unity-gtk-module patch.

**Comment by 1931186** [comment]
Thank you very much, this is exactly what I was looking for [here](https://stackoverflow.com/questions/44143643) I'll have a look into this :)

**Comment by snippins** [comment]
Wow wow thanks to @rilian-la-te, look like we will have a stable HUD for all apps very soon :). Just now lowriter crashed using the HUD before I saw @rilian-la-te comment. Big thank to you guys. See ya.

**Comment by rilian-la-te** [comment]
So, I will only need to reimplement it in normal language? It is a good deal))))

**Comment by snippins** [comment]
@1931186 Ok, I cannot find a way to extract any layout information from Gio.DBusMenuModel apart from creating a menu out of it. However I think the low-level way are still work stable enough for this. Lets hope the DBusActionGroup active() work properly. 

Also noticed Gnome-MPV put actions in the wrong action groups, leads to different object path for action and menu. Yet they still works somehow. I suspect the underlying gtk code trim all the prefixes ("win", "unity", "app") when associate them with the menu items.

**Comment by rilian-la-te** [comment]
But if you just add it to GMenu, can you iterate over attibutes?

**Comment by 1931186** [comment]
@snippins Yeah that's also where I got stuck. Using a gtk menu for this seems a bit unecessary and you are right, retrieving the menu layout seems to be stable enough.

**Comment by snippins** [comment]
@1931186 can you try my current [code](https://gist.github.com/snippins/ee943f2b25db555ef12107f7cee20241) to see if it works ok with GIMP?.

@rilian-la-te cant find any method to do that. Actually DBusMenuModel should have methods for this, but maybe they did not implemented all the interface as @1931186 mentioned (for python atleast).

**Comment by 1931186** [comment]
@snippins It seems to work, the gimp crashes are super inconsistent tho, so I can't tell for sure. I'll have to implement it for the toolbar I'm working on to give some useful feedback. Also if the code you linked works on your machine, you probably have some outdated libraries (your bus_name and menu_object_path are undecoded bytes for me, so if you want this to be portable you should throw some .decode("utf-8")s in there :))

**Comment by snippins** [comment]
@1931186 The python implementation seem to work inconsistent with the specs. I tried to list all actions, its return an empty list, try to check if an entry is enabled, it always return fail. Querying about the action, got false, None, etc ?!? @@! Luckily the active function work...

**Comment by 1931186** [comment]
@snippins Yeah wtf, right? It seems to be the same with GMenuModel. I'm tempted to blame it on pygobject but that's probably not the case. Maybe it's because we ignored some specs regarding retrieving the layout from the bus and now it's not initialized properly? Idk, it's probably not that much of a deal...

**Comment by 1931186** [comment]
Ok, I've found some time to test, this is what I found:
- using DBusActionGroup is stable and works perfectly fine
- calling Start() directly on gtk-menus is extremely unstable and will likely crash more sophisticated applications at some point (especially if you do it directly after some context switching)
- it **seems** like using Gio.DBusMenuModel is stable, but this may be because creating a DBusMenuModel doesn't actually do anything but creating a proxy for later calls. I'll see whether or not I can get python to work with a gtk Menu casted from a DBusMenuModel when I have some time.

@snippins Seems like using Start() for your purpose should work ok, but I wouldn't be too surprised if some applications just crash after some time.

**Comment by snippins** [comment]
Ok I think I got an idea of how to use DBusMenuModel correctly. It has an event call items-changed that need a function to connect to. Then run a Glib mainloop to listen to the events inside. The iterating methods works just find inside these methods. You can check the code [here](https://gist.github.com/snippins/255db0c804a2a81815ad484909ea241c). To use it you can do something like "sleep 3 && thatfile.py" then quicky switch to some gtk app. I am able to retrieve almost everything.

But I am not familiar with this stuff and dont know how to go further yet. How to know when to stop and break the loop.

**Comment by snippins** [comment]
I call End([x for x in range(1024)]) method and it almost never crash anything though.

**Comment by 1931186** [comment]
@snippins  Ohhh that makes sense. You're right, there needs to be a mainloop running for the whole thing to work. I've actually looked into the GLib source code and [DBusMenuModel asks for the main context](https://github.com/GNOME/glib/blob/f924d0b1f7d2b019f1abb56685dcfda74266c608/gio/gdbusmenumodel.c#L886). When looking for the path, it [doesn't find it](https://github.com/GNOME/glib/blob/f924d0b1f7d2b019f1abb56685dcfda74266c608/gio/gdbusmenumodel.c#L321) and creates it. But thats not linked to the actual menu on the bus.

This is also the first time for me having to work with GLib, so I'm also pretty clueless about whether there's a way to bypass this. What I think we have to do is query an arbitrary signal, enter the mainloop, use the signal to trigger menu-discovery, and exit with quit().

What do you mean with 'almost'? Does it crash for you or not? I found that crashes seem to be pretty arbitrary and dependent on the time between opening a new window and running Start(), but when 'stress'-testing, I crash pretty consistantly.

**Comment by snippins** [comment]
@1931186 The problem is how we can know the menu has explored? Of course we can use some timestamps to force quit the loop, but that sound worse than using the dbus interface since no guarantee we got all the layout. 

Not sure if it is related to the crash, but I try to fix the old implementation which does not pass the target to the active function, this lead to some entry not working (GnomeMPV->File->Open). Check my gist and 'stress'-testing your way again. I been testing this one and nothing have crashed yet.

**Comment by 1931186** [comment]
@snippins I don't see why we wouldn't get the whole menu? Doesn't GMenuLinkIter indicate when we reached the end?

The stress-test is really easy, just let `while :; do python HUD.py; done` run in the background after you disabled the rofi part and open some new windows in gimp.

**Comment by snippins** [comment]
@1931186 It works like this: When the first event got triggered <=> the root menu expanded, we use iterate_item_attributes() to get the name and attribute for calling active() later. Then for each item of that we hook the event to a sub item, which is also a GMenuModel. It keeps expanding like and each event is treat as equals, we never know when all of them is done.

I think I know the way to do it know though. The parent event can be block by a loop waiting for its child event to completed, thus back to normal tree traversal.

**Comment by 1931186** [comment]
@snippins Yeah this is more than suboptimal. I think the best solution would be calling Start() one by one starting with [0] to explore all submenus. I can now confirm that the problem is indeed caused by GLib trying to deref non-existing menu-items that got referenced by calling Start(). This will be a more expensive than I would like, but still more sane than running a GLib Mainloop with recursive callbacks.

**Comment by snippins** [comment]
Yeah that way I think the code will be much clearer and more straightforward than using GMenuModel. pydbus is not yet standard so I think we can just stay with the "deprecated" API for now.

**Comment by 1931186** [comment]
@snippins Ok then thats the plan. I'll see if I can implement this tomorrow or so...

Also I think you need to step up your gimp-game lol
![2017-05-27-000456_1366x768_scrot](https://cloud.githubusercontent.com/assets/14981862/26514149/498ceddc-4270-11e7-8def-94c9faa03d21.png)

**Comment by snippins** [comment]
@1931186 Any luck implementing it?

**Comment by 1931186** [comment]
@snippins Yeah, I edited it into the [last post](https://github.com/electron/electron/issues/8297#issuecomment-304397948) above. It works. I also added support for section labels. If you want to use it for your HUD I can comment the code so you can tweak it if you want.

**Comment by snippins** [comment]
@1931186 Thanks a lot. The 'target' is use when calling active(). Most menu entries does not need it but some does. An example is gnome-mpv (File->Open, Edit->Subtitle track->Load external).

**Comment by 1931186** [comment]
@snippins So you want me to comment it? Thanks for the info.

**Comment by snippins** [comment]
@1931186 Yes if you just want to extract the layout. Also do you know which app will fail if the ":section" label does not got supported?

**Comment by 1931186** [comment]
@snippins Ok. No app will fail, it's purely cosmetic.

**Comment by 1931186** [comment]
@snippins ok, I tried my best commenting [it](https://gist.github.com/1931186/caa0ee955300bc2a40ef55d123b06212)

Also bugfixes and implemented the 'target' prop

**Comment by snippins** [comment]
@1931186 Wow thanks for the effort, I will surely try to use it for my HUD.

**Comment by snippins** [comment]
@1931186 Notice that some app also use "win." prefix for its action. I think we should do
action = action.split(".",1)[1] so all the cases are covered.

**Comment by rilian-la-te** [comment]
three prefixes are widely usable:
app, win and unity.

**Comment by snippins** [comment]
@rilian-la-te Thanks.

@1931186 Sadly I have to use the deprecated dbus API, just follow your idea because the target signature is "av" but the target returned by DBus can be anything, thus we need to wrap the "v" part by a Glib.Variant with type "*". However it is not implemented in python yet.

**Comment by 1931186** [comment]
@rilian-la-te Do you know if it is possible for a action to have no prefix at all, e.g. just "New"?

@snippins I thought python-dbus would do this automatically for you because of its [signature guessing](https://dbus.freedesktop.org/doc/dbus-python/doc/tutorial.html#data-types)? It works for me.

**Comment by rilian-la-te** [comment]
AFAIK, we have 3 action namespaces in appmenu, and 3 prefixes. If somethere we have actions without a prefix, I will not take it into account.

**Comment by 1931186** [comment]
@rilian-la-te Ok, thank you very much.

**Comment by snippins** [comment]
@1931186 Yes, that why I still python-dbus instead of GDBus. I merge your implementation into my [HUD](https://gist.github.com/snippins/ee943f2b25db555ef12107f7cee20241) now. You might want to take a look as I might misunderstand something.

**Comment by 1931186** [comment]
@snippins looks good. You might want to take out the reversed() in line 266 because you're sorting your menu anyways, so all it really does is hit the performance.
